gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/SystemTaskUpgrader.java

```diff
@@ -0,0 +1,132 @@
+                            // force the task.operationId to assign the task to the instance
+                            String previousOperationId = task.getOperationId();
+                            task.setOperationId(instanceOperationId);
+                            return processUpgrade(instanceOperationId, task, previousOperationId);
+                        case ONGOING:
+                            // wait until status change
+                            return Single.error(new IllegalStateException("ONGOING task " + taskId + " : trigger a retry"));
+                        default:
+                            // SUCCESS case
+                            return Single.just(true);
+                    }
+                }).retryWhen(new RetryWithDelay(3, 5000)).blockingGet();
+
+        if (!upgraded) {
+            throw getIllegalStateException();
+        }
+
+        return true;
+    }
+
+    protected abstract Single<Boolean> processUpgrade(String instanceOperationId, SystemTask task, String previousOperationId);
+
+    protected abstract IllegalStateException getIllegalStateException();
+
+    protected Single<SystemTask> createSystemTask(String operationId) {
+        final String taskId = getTaskId();
+        SystemTask systemTask = new SystemTask();
+        systemTask.setId(taskId);
+        systemTask.setType(SystemTaskTypes.UPGRADER.name());
+        systemTask.setStatus(SystemTaskStatus.INITIALIZED.name());
+        systemTask.setCreatedAt(new Date());
+        systemTask.setUpdatedAt(systemTask.getCreatedAt());
+        systemTask.setOperationId(operationId);
+        return systemTaskRepository.create(systemTask).onErrorResumeNext(err -> {
+            logger.warn("SystemTask {} can't be created due to '{}'", taskId, err.getMessage());
+            // if the creation fails, try to find the task, this will allow to manage the retry properly
+            return systemTaskRepository.findById(systemTask.getId()).toSingle();
+        });
+    }
+
+
+    protected abstract String getTaskId();
+
+    private class RetryWithDelay implements Function<Flowable<Throwable>, Publisher<?>> {
+        private final int maxRetries;
+        private final int retryDelayMillis;
+        private int retryCount;
+
+        public RetryWithDelay(int retries, int delay) {
+            this.maxRetries = retries;
+            this.retryDelayMillis = delay;
+            this.retryCount = 0;
+        }
+
+        @Override
+        public Publisher<?> apply(@NonNull Flowable<Throwable> attempts) throws Exception {
+            return attempts
+                    .flatMap((throwable) -> {
+                        if (++retryCount < maxRetries) {
+                            // When this Observable calls onNext, the original
+                            // Observable will be retried (i.e. re-subscribed).
+                            return Flowable.timer(retryDelayMillis * (retryCount + 1),
```

Comment:
Missing a cast to long here - I had to look this one up to understand why it's a warning. With the implicit cast, the calculation is done as 32 bit and _then_ cast to a long, so you risk setting the wrong value in some circumstances. I think that's highly unlikely ever to happen here, although it's a good habit to develop.

Reply:
@farmborough indeed.

I am unlikely to touch this since it was just a refactor